metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/ConsensusService.scala

```diff
@@ -461,19 +517,38 @@ object ConsensusService {
     def empty[A <: Agreement] = MessageStash[A](Map.empty)
   }
 
+  /** Count the number of messages received from others in a round,
+    * to determine whether we're out of sync or not in case of a timeout.
+    */
+  case class MessageCounter(
+      past: Int,
```

Comment:
I understand the desire for symmetry, but it doesn't look like `past` is needed at all

Reply:
Yeah, syncing the view number is not something that's covered in the HotStuff paper, so I just came up with something that seemed plausible. Initially I thought we could keep tabs of what view numbers we receive in messages, try to figure it out from those, but then decided to start with this simpler model of just polling the others if we time out. This is rather experimental and subject to change. 

I thought I'd gather the statistic anyway so that at least we can [trace them](https://github.com/input-output-hk/metronome/pull/37/files#diff-6304157ee8ddbbd937b7fb96c7cc976949ff9833f037ee0032780f8d32e3e7b3R232) and see what's happening. While watching the terminals in the robot example trying to get in sync, I actually [modified](https://github.com/input-output-hk/metronome/pull/42/files#diff-6304157ee8ddbbd937b7fb96c7cc976949ff9833f037ee0032780f8d32e3e7b3L259-R284) this logic to make use of the past.

What I realised was that the node with the highest view number can get isolated. Say we start nodes Alice, Bob and Charlie in a staggered fashion, in that order, and by the time Charlie starts their view numbers are 100, 10, 1, respectively. Bob and Charlie will see that they sometimes get messages from the future, when an older node is leading the round, so they initiate view synchronisation. They, however, pick the median view number, which is Bob's, so they converge after a while, but Alice only gets messages from the past, so she never starts the syncing process. 

So I thought: it should be okay for Alice to wind back her view number, if she sees that she's only getting past messages while timing out. 

These heuristics are not infallible, for example if we had 2 groups with identical view numbers within the group, but different between them, they could receive present messages (multiple nodes can think they are leading, one in each group), but I don't have a better idea at the moment, I'm just watching how long it takes to converge. 

I chose the median because I didn't want to trust the highest view, which could be just one Byzantine node trying to lie to us to isolate us from the herd. I thought if the majority of the federation is in sync, then the median is the right number to use for a new node.

Reply:
Thanks for the explanation